discovery/db_discovery.go

```diff
@@ -119,19 +123,84 @@ func cacheDBOrchs(node *core.LivepeerNode, orchs []*lpTypes.Transcoder) ([]*comm
 		return dbOrchs, nil
 	}
 
+	numOrchestrators := len(orchs)
+	ctx, cancel := context.WithTimeout(context.Background(), getOrchestratorsTimeoutLoop)
+	orchChan := make(chan struct{}, numOrchestrators)
+	numResp := 0
+	numSuccessResp := 0
+	respLock := sync.Mutex{}
+
+	getOrchInfo := func(dbOrch *common.DBOrch) {
+		uri := parseURI(dbOrch.ServiceURI)
+		respLock.Lock()
+		defer respLock.Unlock()
+		numResp++
+
+		if uri != nil {
+			info, err := serverGetOrchInfo(ctx, core.NewBroadcaster(node), uri)
+			if err == nil {
+				dbOrch.PricePerPixel, err = common.RatToFixed(big.NewRat(info.PriceInfo.GetPixelsPerUnit(), info.PriceInfo.GetPricePerUnit()), 1000, 3)
+				if err != nil {
+					glog.Error("Error converting orchestrator price per pixel to a fixed point number")
+				} else if err := node.Database.UpdateOrch(dbOrch); err != nil {
+					glog.Error("Error updating Orchestrator in DB: ", err)
+				} else {
+					dbOrchs = append(dbOrchs, dbOrch)
+					numSuccessResp++
+				}
+			} else if monitor.Enabled {
+				monitor.LogDiscoveryError(err.Error())
+			}
+		}
+
+		if numSuccessResp >= numOrchestrators || numResp >= numOrchestrators {
+			orchChan <- struct{}{}
+		}
+	}
+
 	for _, orch := range orchs {
 		if orch == nil {
 			continue
 		}
 		dbOrch := ethOrchToDBOrch(orch)
-		if err := node.Database.UpdateOrch(dbOrch); err != nil {
-			glog.Error("Error updating Orchestrator in DB: ", err)
-			continue
+		go getOrchInfo(dbOrch)
+
+	}
+
+	select {
+	case <-ctx.Done():
+		respLock.Lock()
+		if len(dbOrchs) < numOrchestrators {
+			numOrchestrators = len(dbOrchs)
 		}
-		dbOrchs = append(dbOrchs, dbOrch)
+		returnDbOrchs := dbOrchs[:numOrchestrators]
+		respLock.Unlock()
+		glog.Info("Done fetching orch info for orchestrators, context timeout: ", returnDbOrchs)
+		cancel()
+		return returnDbOrchs, nil
+	case <-orchChan:
+		respLock.Lock()
+		if len(dbOrchs) < numOrchestrators {
+			numOrchestrators = len(dbOrchs)
+		}
+		returnDbOrchs := dbOrchs[:numOrchestrators]
+		respLock.Unlock()
+		glog.Info("Done fetching orch info for orchestrators, numResponses fetched: ", returnDbOrchs)
+		cancel()
+		return returnDbOrchs, nil
 	}
-	return dbOrchs, nil
+}
 
+func parseURI(addr string) *url.URL {
+	if !strings.HasPrefix(addr, "http") {
+		addr = "https://" + addr
```

Comment:
I think so, because we're fetching these from on-chain and bad URL's can thus happen.

Reply:
I'm a bit uneasy about this for a few (admittedly abstract) reasons.

If there's bad data on the blockchain, then there's bad data on the blockchain. If the O registers itself normally (eg, via CLI) then the URL will always have the https prefix. I'm nervous about things interacting unexpectedly [1] or downstream ecosystem impacts [2] from special-cases like these if we take the default behavior of "fixing" URLs. The brevity of not needing a helper function is also nice... Anyway, I'll leave this one up to you.

[1] Can't think of a concrete case right now, but what makes me nervous is this doesn't prove the absence of such cases.

[2] An example: third-party transcoder registration software starts omitting the prefix, which may break other implementations of the Livepeer node (even though there aren't any right now), and forces subsequent implementations to accommodate this special case; implicitly becoming a required part of the "contract" for the ServiceURI

Reply:
Hmm I guess the same functionality here 

https://github.com/livepeer/go-livepeer/blob/17701ca9a7bfb8b84de08da9fae1223ae2053b29/discovery/discovery.go#L35

is then rather to force CLI supplied orchestrator URIs to have the prefix in offchain mode, not per se URIs fetched on chain ?

Reply:
Ah yeah, you're right - IIRC it was meant as a UX affordance and probably belongs better where we handle the CLI flags in [cmd/livepeer.go](https://github.com/livepeer/go-livepeer/blob/17701ca9a7bfb8b84de08da9fae1223ae2053b29/cmd/livepeer/livepeer.go#L157-L165)

Still up to you - especially since we already have this fixup another place, I'm slightly less uneasy (but still a little bit).

Reply:
I left them in place for now. But could still remove this after re-review.